monix-execution/shared/src/main/scala/monix/execution/FutureUtils.scala

```diff
@@ -115,18 +116,28 @@ object FutureUtils {
     *
     * Similar to `Future.transformWith` from Scala 2.12.
     */
-  def transformWith[A,B](source: Future[A], f: Try[A] => Future[B])(implicit ec: ExecutionContext): Future[B] = {
+  def transformWith[A,B](source: Future[A], f: Try[A] => Future[B])(implicit ec: ExecutionContext): CancelableFuture[B] = {
     val p = Promise[B]()
+    val cancelable = MultiAssignmentCancelable()
     source.onComplete { result =>
       val fb = try f(result) catch { case NonFatal(t) => Future.failed(t) }
+      fb match {
+        case c: Cancelable => cancelable.orderedUpdate(c, order = 2)
+        case _ =>
+      }
       if (fb eq source)
         p.complete(result.asInstanceOf[Try[B]])
       else fb.value match {
         case Some(value) => p.complete(value)
         case None => p.completeWith(fb)
       }
     }
-    p.future
+
+    source match {
+      case c: Cancelable => cancelable.orderedUpdate(c, order = 1)
```

Comment:
This is correct, but you already have `source` when initializing your `val cancelable` so it might be a better idea to initialize that as `MultiAssignmentCancelable(source)` (thus `source` being an initial underlying reference), because we avoid doing an `Atomic.compareAndSet` that way.

And so instead of the second `orderedUpdate(order = 2)`, you can just use `:=`, because `orderedUpdate` also implies some extra logic.